File: python/mirage/core/ssh/rmdir.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import asyncssh

from mirage.accessor.ssh import SSHAccessor
from mirage.cache.context import invalidate_after_unlink
from mirage.cache.index import NULL_INDEX, IndexCacheStore
from mirage.core.ssh.client import _abs
from mirage.types import PathSpec
from mirage.utils.errors import enoent, enotdir, enotempty


async def _holds_entries(sftp: asyncssh.SFTPClient, remote: str) -> bool:
    """Whether the remote directory still lists children.

    The probe behind the version-3 arm below: it decides between "not
    empty" and every other refusal the one generic code covers. A probe
    that fails is a negative probe, never an error to surface; the
    caller re-raises what the server said.

    Args:
        sftp (asyncssh.SFTPClient): open SFTP session.
        remote (str): absolute remote path of the directory.
    """
    try:
        names = await sftp.listdir(remote)
    except (OSError, asyncssh.Error):
        return False
    return any(name not in (".", "..") for name in names)


async def rmdir(accessor: SSHAccessor,
                path: PathSpec,
                index: IndexCacheStore = NULL_INDEX) -> None:
    """Remove an empty directory over SFTP.

    The server enforces emptiness, so the work here is naming its
    refusal in the same vocabulary every other backend uses. SFTP 3 has
    one generic code for it (``SFTPFailure``, carrying only the server's
    message string), and later protocol versions split
    ``SFTPDirNotEmpty`` out; OpenSSH speaks version 3, so the generic
    code is what a not-empty rmdir actually answers in practice. It also
    covers other refusals, so one listing probe decides instead of a
    blind translation: only a directory that still shows entries
    converts to ENOTEMPTY, anything else keeps the server's own answer.
    Without the conversion the hidden-remnant guard never fires on ssh
    (it keys on the errno), and the raw SFTP failure leaks.

    Args:
        accessor (SSHAccessor): SSH accessor.
        path (PathSpec): directory to remove.
        index (IndexCacheStore): accepted for the rmdir slot's shape;
            unused.
    """
    config = accessor.config
    sftp = await accessor.sftp()
    remote = _abs(config, path.mount_path)
    try:
        await sftp.rmdir(remote)
    except asyncssh.SFTPNoSuchFile as exc:
        raise enoent(path) from exc
    except asyncssh.SFTPDirNotEmpty as exc:
        raise enotempty(path) from exc
    except asyncssh.SFTPNotADirectory as exc:
        raise enotdir(path) from exc
    except asyncssh.SFTPFailure as exc:
        if await _holds_entries(sftp, remote):
            raise enotempty(path) from exc
        raise
    await invalidate_after_unlink(path)

File: typescript/packages/node/src/core/ssh/rmdir.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import { invalidateAfterUnlink } from '@struktoai/mirage-core/cache/context'
import type { PathSpec } from '@struktoai/mirage-core/types'
import { enoent, enotempty } from '@struktoai/mirage-core/utils/errors'
import type { SSHAccessor } from '../../accessor/ssh.ts'
import { isFailure, isNoSuchFile, joinRoot, stripPrefix } from './utils.ts'

// The probe behind the version-3 arm below: whether the directory
// still lists children. A probe that fails is a negative probe, never
// an error to surface; the caller re-raises what the server said.
async function holdsEntries(accessor: SSHAccessor, remote: string): Promise<boolean> {
  const sftp = await accessor.sftp()
  try {
    return await new Promise<boolean>((resolveFn, rejectFn) => {
      sftp.readdir(remote, (err, entries) => {
        if (err !== undefined) {
          rejectFn(err)
          return
        }
        resolveFn(entries.some((e) => e.filename !== '.' && e.filename !== '..'))
      })
    })
  } catch {
    return false
  }
}

// SFTP 3 has one generic code for a not-empty refusal (SSH_FX_FAILURE),
// which is what ssh2 always speaks, so one listing probe decides
// instead of a blind translation: only a directory that still shows
// entries converts to ENOTEMPTY, anything else keeps the server's own
// answer. Without the conversion the hidden-remnant guard never fires
// on ssh (it keys on the errno), and the raw SFTP failure leaks.
export async function rmdir(accessor: SSHAccessor, p: PathSpec): Promise<void> {
  const sftp = await accessor.sftp()
  const virtual = stripPrefix(p)
  const remote = joinRoot(accessor.config.root ?? '/', virtual)
  try {
    await new Promise<void>((resolveFn, rejectFn) => {
      sftp.rmdir(remote, (err) => {
        if (!err) {
          resolveFn()
          return
        }
        if (isNoSuchFile(err)) rejectFn(enoent(p))
        else rejectFn(err)
      })
    })
  } catch (err) {
    if (isFailure(err) && (await holdsEntries(accessor, remote))) {
      throw enotempty(p)
    }
    throw err
  }
  await invalidateAfterUnlink(p)
}
